source/framework/vx_reference.c

```diff
@@ -415,6 +415,136 @@ vx_status ownDestructReferenceGeneric(vx_reference ref)
+        status = ownReferenceLock(output);
+        if (((vx_status)VX_SUCCESS == status) &&
+            (ip_mem_size <= op_mem_size))
+        {
+            status = tivxMemBufferMap((void *)(uintptr_t)ip_mem_ptr->host_ptr, ip_mem_size, (vx_enum)VX_MEMORY_TYPE_HOST, (vx_enum)VX_READ_ONLY);
+            if ((vx_status)VX_SUCCESS == status)
+            {
+                status = tivxMemBufferMap((void *)(uintptr_t)op_mem_ptr->host_ptr, op_mem_size, (vx_enum)VX_MEMORY_TYPE_HOST, (vx_enum)VX_WRITE_ONLY);
+                if ((vx_status)VX_SUCCESS == status)
+                {
+                    memcpy((void *)(uintptr_t)op_mem_ptr->host_ptr, (void *)(uintptr_t)ip_mem_ptr->host_ptr, ip_mem_size);
+                    /* copy specific fields*/
+                    if ((vx_enum)VX_TYPE_ARRAY == input->type)
+                    {
+                        tivx_obj_desc_array_t *ip_obj_desc = (tivx_obj_desc_array_t *)input->obj_desc;
+                        tivx_obj_desc_array_t *op_obj_desc = (tivx_obj_desc_array_t *)output->obj_desc;
+                        op_obj_desc->num_items = ip_obj_desc->num_items;
+                    }
+                    else if ((vx_enum)VX_TYPE_USER_DATA_OBJECT == input->type)
+                    {
+                        tivx_obj_desc_user_data_object_t *ip_obj_desc = (tivx_obj_desc_user_data_object_t *)input->obj_desc;
+                        tivx_obj_desc_user_data_object_t *op_obj_desc = (tivx_obj_desc_user_data_object_t *)output->obj_desc;
+                         op_obj_desc->valid_mem_size = ip_obj_desc->valid_mem_size;
+                    }
+                    status =  tivxMemBufferUnmap((void *)(uintptr_t)op_mem_ptr->host_ptr, op_mem_size, (vx_enum)VX_MEMORY_TYPE_HOST, (vx_enum)VX_WRITE_ONLY);
+                }
+                status =  tivxMemBufferUnmap((void *)(uintptr_t)ip_mem_ptr->host_ptr, ip_mem_size, (vx_enum)VX_MEMORY_TYPE_HOST, (vx_enum)VX_READ_ONLY);
+            }
+        }
+        status = ownReferenceUnlock(output);
+    }
+    else
+    {
+        status = (vx_status)VX_FAILURE;
+    }
+
+    return status;
+}
+
+vx_status ownSwapReferenceGeneric(vx_reference input, vx_reference output)
+{
+    vx_status status = (vx_status)VX_FAILURE;
+    tivx_shared_mem_ptr_t  *ip_mem_ptr = NULL;
+    volatile uint32_t       ip_mem_size = 0;
+    tivx_shared_mem_ptr_t  *op_mem_ptr = NULL;
+    volatile uint32_t       op_mem_size = 0;
+    
+    if((vx_bool)vx_true_e == ownIsGenericAllocReferenceType(input->type) &&
+       (vx_bool)vx_true_e == ownIsGenericAllocReferenceType(output->type))
+    {
+        /* added void here as this status check of
+        * ownReferenceGetMemAttrsFromObjDesc will always be true if
+        * the previous condition is true
+        */
+        (void)ownReferenceGetMemAttrsFromObjDesc(input, &ip_mem_ptr, &ip_mem_size);
+        (void)ownReferenceGetMemAttrsFromObjDesc(output, &op_mem_ptr, &op_mem_size);
+
+        status =  ownReferenceLock(output);
```

Comment:
This should lock both the input and output, correct?

Reply:
done